crates/core/tedge_api/src/alarm.rs

```diff
@@ -1,58 +1,84 @@
-use std::convert::TryFrom;
-use std::convert::TryInto;
-use std::fmt;
-
+use crate::entity_store::EntityMetadata;
 use clock::Timestamp;
 use serde::Deserialize;
 use serde::Serialize;
-
 use serde_json::Value;
 use std::collections::HashMap;
-/// In-memory representation of ThinEdge JSON alarm.
-#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+use std::convert::TryFrom;
+
+/// Internal representation of ThinEdge alarm model.
+#[derive(Debug, Eq, PartialEq)]
 pub struct ThinEdgeAlarm {
-    pub name: String,
-    pub severity: AlarmSeverity,
-    #[serde(flatten)]
-    pub data: Option<ThinEdgeAlarmData>,
-    pub source: Option<String>,
-}
+    /// Alarm type retrieved from the MQTT topic or payload.
+    /// The "type" given from the payload has priority to the one from the topic.
+    pub alarm_type: Option<String>,
 
-#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
-pub enum AlarmSeverity {
-    Critical,
-    Major,
-    Minor,
-    Warning,
+    /// Alarm source retrieved from the MQTT topic.
+    pub source: EntityMetadata,
+
+    /// All other alarm data retrieved from the MQTT payload.
+    /// None means the message is meant to clear an alarm.
+    pub data: Option<ThinEdgeAlarmData>,
 }
 
-/// In-memory representation of ThinEdge JSON alarm payload
+/// Internal representation of the JSON MQTT payload.
 #[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
 pub struct ThinEdgeAlarmData {
+    pub severity: Option<String>,
+
     pub text: Option<String>,
 
     #[serde(default)]
     #[serde(with = "time::serde::rfc3339::option")]
     pub time: Option<Timestamp>,
 
+    #[serde(default)]
     #[serde(flatten)]
-    pub alarm_data: HashMap<String, Value>,
+    pub extras: HashMap<String, Value>,
 }
 
-#[derive(thiserror::Error, Debug)]
-pub enum ThinEdgeJsonDeserializerError {
-    #[error("Unsupported topic: {0}")]
-    UnsupportedTopic(String),
+impl ThinEdgeAlarm {
+    pub fn new(alarm_type: &str, source: &EntityMetadata) -> Self {
+        Self {
+            alarm_type: Some(alarm_type.into()),
+            source: source.clone(),
+            data: None,
+        }
+    }
 
-    #[error("Unsupported alarm severity in topic: {0}")]
-    UnsupportedAlarmSeverity(String),
+    /// Override "type" when it is provided in payload.
+    pub fn with_alarm_data(self, mut data: ThinEdgeAlarmData) -> Self {
+        match data.extras.remove("type") {
+            Some(maybe_type_from_payload) => match maybe_type_from_payload.as_str() {
```

Comment:
Why do we need this second match ` maybe_type_from_payload.as_str()` ?

Reply:
The hashmap `data.extras` may have `type: true` (boolean), not the desirable type. In this case, `as_str()` returns `Err`. If it's `Err`, I want to use the default type `ThinEdgeAlarm`.